snapcraft/tests/fake_servers.py

```diff
@@ -0,0 +1,277 @@
+                    self.path))
+            raise NotImplementedError(self.path)
+
+    def _handle_scan_complete_request(self):
+        logger.debug('Handling scan complete request')
+        self.send_response(200)
+        self.send_header('Content-Type', 'application/json')
+        self.end_headers()
+        response = {
+             'message': '',
+             'application_url': 'test-application-url',
+             'revision': 'test-revision',
+             'completed': True
+        }
+        self.wfile.write(json.dumps(response).encode())
+
+
+class FakeStoreSearchServer(http.server.HTTPServer):
+
+    def __init__(self, server_address):
+        super().__init__(
+            server_address, FakeStoreSearchRequestHandler)
+
+
+class FakeStoreSearchRequestHandler(http.server.BaseHTTPRequestHandler):
+
+    # XXX This fake server as reused as download server, to avoid passing a
+    # port as an argument. --elopio - 2016-05-01
+
+    _API_PATH = '/api/v1/'
+
+    def do_GET(self):
+        parsed_path = urllib.parse.urlparse(self.path)
+        search_path = self._API_PATH + 'search'
+        download_path = '/download-snap/'
+        if parsed_path.path.startswith(search_path):
+            self._handle_search_request(
+                urllib.parse.parse_qs(parsed_path.query)['q'])
+        elif parsed_path.path.startswith(download_path):
+            self._handle_download_request(parsed_path[len(download_path):])
+        else:
+            logger.error(
+                'Not implemented path in fake Store Search server: {}'.format(
+                    self.path))
+            raise NotImplementedError(self.path)
+
+    def _handle_search_request(self, query):
+        if len(query) > 1 or 'package_name:' not in query[0]:
+            logger.error(
+                'Not implemented query in fake Store Search server: {}'.format(
+                    query))
+            raise NotImplementedError(query)
+        query = query[0]
+        package = query.split('package_name:')[1].strip('"')
+        logger.debug(
+            'Handling search request for package {}, with headers {}'.format(
+                package, self.headers))
+        self.send_response(200)
+        self.send_header('Content-Type', 'application/hal+json')
+        self.end_headers()
+        if package == 'test-snap':
+            response = {'_embedded': {
+                'clickindex:package': [
+                    {'download_url': urllib.parse.urljoin(
+                        'http://localhost:{}'.format(self.server.server_port),
+                        'download-snap/test-snap.snap'),
+                     # sha512sum snapcraft/tests/data/test-snap.snap
+                     'download_sha512': (
+                        '69d57dcacf4f126592d4e6ff689ad8bb8a083c7b9fe44f6e738ef'
+                        'd22a956457f14146f7f067b47bd976cf0292f2993ad864ccb498b'
+                        'fda4128234e4c201f28fe9')}]}}
```

Comment:
Ha ! Good one ;)
The macaroons tapes could be made smaller by using that, the key point here is indeed to keep a snap and its matching download_sha512 in sync.